framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -2429,68 +2429,79 @@ private static final boolean validType(TypeMirror type) {
         if (input != null) {
             StubParser stubParser = new StubParser("flow.astub", input, this, processingEnv);
-            stubParser.parse(indexTypes, indexDeclAnnos);
+            stubParser.parse(typesFromStubFiles, declAnnosFromStubFiles);
         }
 
-        String allstubFiles = "";
-        String stubFiles;
+        // Stub files specified via stubs compiler option, stubs system property,
+        // stubs env. variable, or @Stubfiles
+        String allStubFiles = "";
 
-        stubFiles = checker.getOption("stubs");
-        if (stubFiles != null)
-            allstubFiles += File.pathSeparator + stubFiles;
+        // 3. Stub files provide via stubs option
+        String stubsOption = checker.getOption("stubs");
+        allStubFiles += stubsOption != null ? File.pathSeparator + stubsOption : "";
 
-        stubFiles = System.getProperty("stubs");
-        if (stubFiles != null)
-            allstubFiles += File.pathSeparator + stubFiles;
+        // 4. Stub files provide via stubs system property
+        String stubsProperty = System.getProperty("stubs");
+        allStubFiles += stubsProperty != null ? File.pathSeparator + stubsProperty : "";
 
-        stubFiles = System.getenv("stubs");
-        if (stubFiles != null)
-            allstubFiles += File.pathSeparator + stubFiles;
+        // 5. Stub files provide via stubs environment variable
+        String stubEnvVar = System.getenv("stubs");
+        allStubFiles += stubEnvVar != null ? File.pathSeparator + stubEnvVar : "";
 
-        {
-            StubFiles sfanno = checker.getClass().getAnnotation(StubFiles.class);
-            if (sfanno != null) {
-                String[] sfarr = sfanno.value();
-                stubFiles = "";
-                for (String sf : sfarr) {
-                    stubFiles += File.pathSeparator + sf;
-                }
-                allstubFiles += stubFiles;
+        // 6. Stub files listed in @Stubfiles annotation on the checker
+        StubFiles stubFilesAnnotation = checker.getClass().getAnnotation(StubFiles.class);
+        if (stubFilesAnnotation != null) {
+            String[] sfarr = stubFilesAnnotation.value();
+            for (String sf : sfarr) {
+                allStubFiles += File.pathSeparator + sf;
             }
         }
 
-        if (allstubFiles.isEmpty()) {
-            this.indexTypes = indexTypes;
-            this.indexDeclAnnos = indexDeclAnnos;
+        if (allStubFiles.isEmpty()) {
+            this.typesFromStubFiles = typesFromStubFiles;
+            this.declAnnosFromStubFiles = declAnnosFromStubFiles;
             return;
         }
 
-        String[] stubArray = allstubFiles.split(File.pathSeparator);
+        // Parse stub files specified via stubs compiler option, stubs system property,
+        // stubs env. variable, or @Stubfiles
+        String[] stubArray = allStubFiles.split(File.pathSeparator);
```

Comment:
Instead of first building up a long String and splitting it up again, would it be easier to use a list?